tests/rgb_tests.cpp

```diff
@@ -131,3 +134,103 @@ TEST(rgb, linear_to_srgb8_lut)
 		EXPECT_EQ(linear_to_srgb8_lut(lin), expected);
 	}
 }
+
+TEST(rgb, rgb565_in_memory_order)
+{
+	//                                  | red | green| blue|
+	constexpr uint16_t rgb_as_uint16 = 0b00001'000011'00111;
+	constexpr uint8_t r8             = 0b00001'000;
+	constexpr uint8_t g8             = /* */ 0b000011'00;
+	constexpr uint8_t b8             = /*        */ 0b00111'000;
+
+	const Rgb565 rgb_object_from_uint16(rgb_as_uint16);
+	const Rgb565 rgb_object_from_components(r8, g8, b8);
+
+	EXPECT_EQ(rgb_object_from_uint16.pixel, rgb_object_from_components.pixel);
+}
+
+TEST(rgb, rgb666_in_memory_order)
+{
+	constexpr uint8_t r6 = 0b100'011;
+	constexpr uint8_t g6 = 0b100'111;
+	constexpr uint8_t b6 = 0b101'111;
+
+	constexpr Rgb666 rgb_object(r6, g6, b6);
+
+	constexpr auto array_size = sizeof(rgb_object);
+
+	// Copy the object byte-by-byte into an array
+	uint8_t rgb_array[array_size] = {};
+	memcpy(rgb_array, reinterpret_cast<const void*>(&rgb_object), array_size);
```

Comment:
I'm a bit uncomfortable about this. The RGB components are just defined as member vars in the class:

```cpp
public:
	uint8_t red   = 0;
	uint8_t green = 0;
	uint8_t blue  = 0;
```

I'd prefer if you converted the class to the `union` / `struct` trick you used in `bgra8888.h`, otherwise this seems a bit hacky to me to use `rgb666` this way.

Plus we'd need some explanatory class comments in `rgb666`, similarly to `bgra8888.h`.

But taking a step back, aren't we always using RGB666 and RGB565 packed to `uint16_t` in the actual code? If not, the proposed change isn't needed, and I'd get rid of the test testing the memory layout.

Reply:
This chunk of code is only in the unit tests to see what colour values we get if the memory is copied as a byte-ordered array.  

Basically to check the underlying representation.

But I can avoid the cast by using an array, and will inspect some byte-offsets as another approach.

> I'd prefer if you converted the class to the union / struct trick you used in bgra8888.h

Unfortunately the Rgb888 object needs to be 3-bytes packed, so a union with a `uint32_t` would push the object's size up to 32-bit (and that `int()` conversion cast is used heavily in the render.h routines.. so going to avoid touching this header for now :-).

Reply:
Cool, but maybe I wasn't entirely clear 😄 `Rgb666` is only used for the palettes, so I guess I'm questioning if need to _care_ about the actual representation at all. In other words, if we always read/write `Rgb666` entries by their component values, then it does not matter one bit what the underlying representation is.

My view is that we should avoid making assumption about the underlying representation and such casting unless we absolutely must.

It makes sense for `Bgra8888` you introduced because the purpose of that is really to serve as an interface/view to the underlying representation. Same for `Rgb555` and `Rgb565` as well that function as a "view" into an `uint16_t` with a specific bit-layout.

I'm just questioning if we need all that for `Rgb666`. Can't we just leave it abstract and make no assumptions on the underlying representation if it's only used for the palette stuff and the individual r6, g6, b6 values get mapped to 8-bit values on an individual component basis anyway (I'm pretty sure that's only how it's used; I'd be surprised if any code treated `Rgb666` values a "array of bytes" in the code).